### PR TITLE
chg: [var] Added variables, re-order user-data and fixed time-out

### PR DESCRIPTION
**Pull request by SteveClement**
- ssh would timeout so I added a more lenient waiting time/handshake
- in user-data password came before user, which is not logical IMHO
- added variables for the username and password
- added some more documentation in case of similar failures
- added **shutdown_command**

### CODE REVIEW

**Comment by SteveClement** [comment]
My build with VMware WS 15.5 takes around 8minutes...

PACKER_LOG=1 packer build ubuntu-2004.json  8.78s user 7.87s system 3% cpu 8:18.39 total

File: ubuntu-2004.json [hunk 1]
```diff
@@ -10,14 +10,17 @@
       "iso_checksum": "caf3fd69c77c439f162e2ba6040e9c320c4ff0d69aad1340a514319a9264df9f",
       "iso_checksum_type": "sha256",
 
-      "ssh_username": "ubuntu",
-      "ssh_password": "ubuntu",
-      "ssh_handshake_attempts": "20",
+      "ssh_username": "{{user `ssh_user`}}",
+      "ssh_password": "{{user `ssh_pass`}}",
+      "ssh_handshake_attempts": "30",
+      "ssh_timeout": "10m",
```
Comment (nickcharlton) [inline ubuntu-2004.json:16]: I was curious that we'd need to do this, but I notice the wording is different in the docs:

> `ssh_timeout` (duration string | ex: "1h5m2s") - The time to wait for SSH to become available. Packer uses this to determine when the machine has booted so this is usually quite long. Example value: 10m.

https://www.packer.io/docs/builders/vmware/iso/#ssh_timeout

Are you finding that there's a lot of failures without specifying this?

File: ubuntu-2004.json [hunk 2]
```diff
@@ -32,5 +35,9 @@
       "type": "shell",
       "inline": ["ls /"]
     }
-  ]
+  ],
+  "variables": {
+    "ssh_user": "ubuntu",
+    "ssh_pass": "ubuntu"
+  }
```
Comment (nickcharlton) [inline ubuntu-2004.json:42]: Could we move the `variables` block to the top of the file? My thinking is that if you were to read from the top to bottom, you'd come across the first variable reference and then have to search the full file to see where it was defined.

File: ubuntu-2004.json [hunk 3]
```diff
@@ -10,14 +10,17 @@
       "iso_checksum": "caf3fd69c77c439f162e2ba6040e9c320c4ff0d69aad1340a514319a9264df9f",
       "iso_checksum_type": "sha256",
 
-      "ssh_username": "ubuntu",
-      "ssh_password": "ubuntu",
-      "ssh_handshake_attempts": "20",
+      "ssh_username": "{{user `ssh_user`}}",
+      "ssh_password": "{{user `ssh_pass`}}",
+      "ssh_handshake_attempts": "30",
```
Comment (nickcharlton) [inline ubuntu-2004.json:15]: This is of course just an arbitrary number I found worked reasonably well, but were you finding it was often hitting more than 20 failures?

**Comment by mweimerskirch** [comment]
Instead of increasing the number of handshake attempts, I just disabled that first SSH server which IMHO is a cleaner solution. See #2

**Comment by nickcharlton** [comment]
Thanks for opening this. I just merged #2, which is a clearer solution to one of the problems address here but I feel like the rest are things people might wish to do in their own solutions — this repo is here as an example and I expect people to take what's here and do their own things with it.

**Pull request closed without merging**